### PR TITLE
Remove workarounds for several compiler bugs

### PR DESCRIPTION
**Pull request by CaseyCarter**
Removes workarounds for:
* VSO-938163 "`constexpr` hates pointers", revealing VSO-1132704 "Bogus C3615 when implicitly-constexpr defaulted SMF calls non-trivial SMF in base class" and VSO-1217687 "Casting object reference to base and back loses track of member values in constant expression" beneath.
* VSO-1002863 "Failure to short-circuit substitution when evaluating disjunction of atomic constraints"
* VSO-1062618 "Asymmetric behavior from conditional operator with char32_t operands"
* VSO-1125263 "EDG rejects valid code using consteval"
* VSO-1132105 "[EDG] Error using non-static data member with dependent type in noexcept-specifier of member function template"
* VSO-1132186 "[EDG] Weirdness with dependent template in noexcept-specifier"
* VSO-1178211 "[EDG] Internal compiler error on template template parameter instantiated from inherited template"

Also relabels the workarounds for VSO-1174090 "Assertion failure for member subobject with multiple prospective destructors" (which is fixed!) as workarounds for VSO-1225825 "ModuleReader doesn't understand concept overloading". 

As a bonus, remove unnecessary workarounds for DevCom-1159442 "type-constraint in deduced variable type fails when immediately following an if statement"; these type-constraints don't follow the _compound-statement_ of an `if` conditional, and so don't trigger the bug.

### CODE REVIEW

**Review by miscco — APPROVED** [review]
All those red lines :+1:

**Review by StephanTLavavej — APPROVED** [review]

**Comment by fsb4000** [comment]
great, nice work compiler team and @CaseyCarter  👍

**Review by cbezault — APPROVED** [review]

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]
This caused the MSVC-internal modules integration tests to fail, so we've had to drop it from the batch of PRs being merged. (I've verified that the modules integration tests are passing without this.) We need to investigate this.

File: stl/inc/iterator [hunk 1]
```diff
@@ -601,16 +601,16 @@ public:
...
     }
 
-#if 0 // TRANSITION, VSO-1174090
+#if 0 // TRANSITION, VSO-1225825
     // clang-format off
     ~_Variantish() requires is_trivially_destructible_v<_It> && is_trivially_destructible_v<_Se> = default;
     // clang-format on
-#endif // TRANSITION, VSO-1174090
+#endif // TRANSITION, VSO-1225825
     ~_Variantish() {
```
Comment (StephanTLavavej) [inline stl/inc/iterator:609]: Should you preserve the newline you were adding here earlier?

**Review by StephanTLavavej — APPROVED** [review]
Looks good to me - one newline comment but you don't need to reset testing for it (it can be fixed when the workaround is finally removed).

File: stl/inc/iterator (hunk 1 above)
Comment (CaseyCarter) [reply stl/inc/iterator]: Yes, that was a righteous style change. I erred on the side of backing everything out here - I'll restore the newline.

**Comment by CaseyCarter** [comment]
Thanks for this compiler regression test coverage!

**Pull request merged**